storage/replica.go

```diff
@@ -341,11 +341,27 @@ func (r *Replica) requestLeaderLease(timestamp roachpb.Timestamp) error {
 	// Note that the command itself isn't traced, but usually the caller
 	// waiting for the result has an active Trace.
 	errChan, pendingCmd := r.proposeRaftCommand(r.context(), ba)
-	if err := <-errChan; err != nil {
-		return err
+
+	// It does no good to block here for longer than the proposed lease's
+	// lifetime. If the command takes that long to commit we are probably
+	// partitioned away and we can conclude that another node has taken
+	// the lease.
+	deadline := time.After(duration)
```

Comment:
Ah, that makes sense. It would be nice to be able to handle a cancelled Context at the lower level so a partitioned node doesn't try to replay all its queued lease requests once it gets reconnected. I'll add a deadline to `r.context()` before the call to `r.proposeRaftCommand`.